File: p21/p2130/PairSum.java
package leetcode.p21.p2130;

import java.util.ArrayList;
import java.util.List;

class ListNode {
    int val;
    ListNode next;

    ListNode() {
    }

    ListNode(int val) {
        this.val = val;
    }

    ListNode(int val, ListNode next) {
        this.val = val;
        this.next = next;
    }
}

public class PairSum {
    /**
     * Mar 06, 2024 22:07
     * Slow Fast Pointer
     * Time Complexity: O(n)
     * Space Complexity: O(n)
     */
    public int pairSum(ListNode head) {
        ListNode slow = head;
        ListNode fast = head;
        List<Integer> pairSums = new ArrayList<>();
        while (fast != null && fast.next != null) {
            pairSums.add(slow.val);
            slow = slow.next;
            fast = fast.next.next;
        }
        int i = pairSums.size() - 1;
        int max = Integer.MIN_VALUE;
        while (slow != null) {
            max = Math.max(max, pairSums.get(i) + slow.val);
            i--;
            slow = slow.next;
        }
        return max;
    }
}

File: p21/p2130/pair_sum.py
# Definition for singly-linked list.
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    """
    Dec 24, 2024 14:41
    Time Complexity: O(n)
    Space Complexity: O(n)
    """

    def pairSum(self, head: Optional[ListNode]) -> int:
        pair_sum = []
        slow = fast = head
        pair_sum.append(slow.val)
        while fast.next and fast.next.next:
            slow = slow.next
            fast = fast.next.next
            pair_sum.append(slow.val)

        i = len(pair_sum) - 1
        max_val = 0
        while slow.next:
            slow = slow.next
            pair_sum[i] += slow.val
            max_val = max(max_val, pair_sum[i])
            i -= 1

        return max_val


if __name__ == '__main__':
    sol = Solution()

    head = ListNode(5, ListNode(4, ListNode(3, ListNode(2))))

    sol.pairSum(head)
